File: 3558-number-of-ways-to-assign-edge-weights-i/3558-number-of-ways-to-assign-edge-weights-i.java
class Solution {
    Map<Integer, List<Integer>> graph = new HashMap<>();
    Set<Integer> visited = new HashSet<>();

    public int assignEdgeWeights(int[][] edges) {
        for (int[] edge : edges){
            int u = edge[0];
            int v = edge[1];

            graph.computeIfAbsent(u,k -> new ArrayList<>()).add(v);
            graph.computeIfAbsent(v,k -> new ArrayList<>()).add(u);   
        }

        int max_depth = dfs(1);
        int MOD = 1_000_000_007;
        return (int) modPow(2, max_depth - 1, MOD);
    }

    public int dfs(int node){
        visited.add(node);
        
        int curr_depth = 0;

        for (int nei : graph.get(node)){
            if (!visited.contains(nei)){
                curr_depth = Math.max(curr_depth, 1 + dfs(nei));
            }
        }

        return curr_depth;
    }

    private long modPow(long base, long exp, long mod) {
        long result = 1;

        while (exp > 0) {
            if ((exp & 1) == 1) {
                result = (result * base) % mod;
            }
            base = (base * base) % mod;
            exp >>= 1;
        }

        return result;
    }
}

File: 3558-number-of-ways-to-assign-edge-weights-i/3558-number-of-ways-to-assign-edge-weights-i.py
class Solution:
    def assignEdgeWeights(self, edges: List[List[int]]) -> int:
        graph = defaultdict(list)

        for u,v in edges:
            graph[u].append(v)
            graph[v].append(u)

        visited = set()

        def dfs(node):
            if not graph[node]:
                return 0

            visited.add(node)
            curr_depth = 0
            for nei in graph[node]:
                if nei not in visited:
                    curr_depth = max(curr_depth, 1 + dfs(nei))

            return curr_depth

        max_depth = dfs(1)

        return (2 ** (max_depth - 1)) % (10**9 + 7)


        